File: android-beans/src/test/java/rocks/frieler/android/beans/scopes/GenericScopedFactoryBeanTest.java
package rocks.frieler.android.beans.scopes;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.junit.MockitoJUnitRunner;

import java8.util.function.Supplier;

import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class GenericScopedFactoryBeanTest {

    private static final String SCOPE = "aScope";
    @SuppressWarnings("unchecked")
    private Supplier<GenericScopedFactoryBeanTest> producer = mock(Supplier.class);
    private GenericScopedFactoryBean<GenericScopedFactoryBeanTest> factoryBean = new GenericScopedFactoryBean<>(SCOPE, GenericScopedFactoryBeanTest.class, producer);

    @Test
    public void testGetScopeReturnsConfiguredScope() {
        assertThat(factoryBean.getScope(), is(SCOPE));
    }

    @Test
    public void testGetTypeReturnsConfiguredType() {
        assertThat(factoryBean.getBeanType(), is(equalTo(GenericScopedFactoryBeanTest.class)));
    }

    @Test
    public void testProduceBeanCallsConfiguredProducer() {
        when(producer.get()).thenReturn(this);

        assertThat(factoryBean.produceBean(), is(this));
    }
}

File: android-beans/src/test/kotlin/rocks/frieler/android/beans/scopes/GenericScopedFactoryBeanTest.kt
package rocks.frieler.android.beans.scopes

import assertk.assertThat
import assertk.assertions.isEqualTo
import assertk.assertions.isSameAs
import assertk.assertions.isSameInstanceAs
import org.junit.jupiter.api.Test
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import rocks.frieler.android.beans.BeansProvider


class GenericScopedFactoryBeanTest {
    private val scope = "aScope"
    private val producer: BeansProvider.() -> GenericScopedFactoryBeanTest = mock()
    private val factoryBean = GenericScopedFactoryBean(scope, GenericScopedFactoryBeanTest::class, producer)

    @Test
    fun `getScope() returns configured scope`() {
        assertThat(factoryBean.scope).isEqualTo(scope)
    }

    @Test
    fun `getType() returns configured type`() {
        assertThat(factoryBean.beanType).isEqualTo(GenericScopedFactoryBeanTest::class)
    }

    @Test
    fun `produceBean() calls the configured producer`() {
        val beansProvider: BeansProvider = mock()
        whenever(producer(beansProvider)).thenReturn(this)

        val producedBean = factoryBean.produceBean(beansProvider)

        verify(producer).invoke(beansProvider)
        assertThat(producedBean).isSameInstanceAs(this)
    }
}
